internal/service/vpc/resource_azure_vpc_peering_connection.go

```diff
@@ -0,0 +1,363 @@
+	},
+	"peering_connection_id": {
+		Computed:    true,
+		Type:        schema.TypeString,
+		Description: "Cloud provider identifier for the peering connection if available",
+	},
+	"peer_azure_app_id": {
+		Required:    true,
+		ForceNew:    true,
+		Type:        schema.TypeString,
+		Description: schemautil.Complex("Azure app registration id in UUID4 form that is allowed to create a peering to the peer vnet").ForceNew().Build(),
+	},
+	"peer_azure_tenant_id": {
+		Required:    true,
+		ForceNew:    true,
+		Type:        schema.TypeString,
+		Description: schemautil.Complex("Azure tenant id in UUID4 form.").ForceNew().Build(),
+	},
+}
+
+func ResourceAzureVPCPeeringConnection() *schema.Resource {
+	return &schema.Resource{
+		Description:   "The Azure VPC Peering Connection resource allows the creation and management of Aiven VPC Peering Connections.",
+		CreateContext: resourceAzureVPCPeeringConnectionCreate,
+		ReadContext:   resourceAzureVPCPeeringConnectionRead,
+		DeleteContext: resourceAzureVPCPeeringConnectionDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceAzureVPCPeeringConnectionImport,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(2 * time.Minute),
+			Delete: schema.DefaultTimeout(2 * time.Minute),
+		},
+
+		Schema: aivenAzureVPCPeeringConnectionSchema,
+	}
+}
+
+func resourceAzureVPCPeeringConnectionCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var (
+		pc     *aiven.VPCPeeringConnection
+		err    error
+		region *string
+	)
+
+	client := m.(*aiven.Client)
+	projectName, vpcID := schemautil.SplitResourceID2(d.Get("vpc_id").(string))
+	if projectName == "" || vpcID == "" {
+		return diag.Errorf("incorrect VPC ID, expected structure <PROJECT_NAME>/<VPC_ID>")
+	}
+
+	azureSubscriptionId := d.Get("azure_subscription_id").(string)
+	vnetName := d.Get("vnet_name").(string)
+
+	// Azure related fields are only available for VPC Peering Connection resource but
+	// not for Transit Gateway VPC Attachment therefore ResourceData.Get retrieves nil
+	// for fields that are not present in the schema.
+	var peerAzureAppID, peerAzureTenantID, peerResourceGroup string
+	if v, ok := d.GetOk("peer_azure_app_id"); ok {
+		peerAzureAppID = v.(string)
+	}
+	if v, ok := d.GetOk("peer_azure_tenant_id"); ok {
+		peerAzureTenantID = v.(string)
+	}
+	if v, ok := d.GetOk("peer_resource_group"); ok {
+		peerResourceGroup = v.(string)
+	}
+
+	if _, err = client.VPCPeeringConnections.Create(
+		projectName,
+		vpcID,
+		aiven.CreateVPCPeeringConnectionRequest{
+			PeerCloudAccount:  azureSubscriptionId,
+			PeerVPC:           vnetName,
+			PeerRegion:        region,
+			PeerAzureAppId:    peerAzureAppID,
+			PeerAzureTenantId: peerAzureTenantID,
+			PeerResourceGroup: peerResourceGroup,
+		},
+	); err != nil {
+		return diag.Errorf("Error waiting for VPC peering connection creation: %s", err)
+	}
+
+	stateChangeConf := &resource.StateChangeConf{
```

Comment:
The state change configuration probably can be de-duplicated between all clouds.

Reply:
In the current shape yes it could be a good candidate for de-duplication, however, we have a problem (not related to this PR) with this part of the code that will be fixed in a separate PR. We are trying to get Azure, GCP and AWS VPC peering connection using the same set of fields which is not enough and it only works properly for the AWS now and in some cases for GCP. So this logic will be updated soon and we will have a different state change waiter and get functions for each cloud type.